contrib/envoyproxy/envoy/fakehttp.go

```diff
@@ -0,0 +1,189 @@
+	// https://datadoghq.atlassian.net/wiki/spaces/TS/pages/2766733526/Sensitive+IP+information#Where-does-the-value-of-the-http.client_ip-tag-come-from%3F
+	return xfwd[length-1]
+}
+
+// partitionPeusdoHeaders Separate normal headers of the initial request made by the client and the pseudo headers of HTTP/2
+// - Format the headers to be used by the tracer as a map[string][]string
+// - Set headers keys to be canonical
+func partitionPeusdoHeaders(receivedHeaders []*corev3.HeaderValue) (map[string][]string, map[string]string) {
+	headers := make(map[string][]string, len(receivedHeaders)-4)
+	pseudoHeaders := make(map[string]string, 4)
+	for _, v := range receivedHeaders {
+		key := v.GetKey()
+		if key == "" {
+			continue
+		}
+		if key[0] == ':' {
+			pseudoHeaders[key] = string(v.GetRawValue())
+			continue
+		}
+
+		headers[http.CanonicalHeaderKey(key)] = []string{string(v.GetRawValue())}
+	}
+	return headers, pseudoHeaders
+}
+
+func NewFakeResponseWriterFromExtProc(w http.ResponseWriter, res *extproc.ProcessingRequest_ResponseHeaders) error {
+	headers, pseudoHeaders := partitionPeusdoHeaders(res.ResponseHeaders.GetHeaders().GetHeaders())
+
+	if err := checkPseudoResponseHeaders(pseudoHeaders); err != nil {
+		return err
+	}
+
+	status, err := strconv.Atoi(pseudoHeaders[":status"])
+	if err != nil {
+		return fmt.Errorf("error parsing status code %q: %w", pseudoHeaders[":status"], err)
+	}
+
+	for k, v := range headers {
+		w.Header().Set(k, strings.Join(v, ","))
+	}
+
+	w.WriteHeader(status)
+	return nil
+}
+
+// NewRequestFromExtProc creates a new http.Request from an ext_proc RequestHeaders message
+func NewRequestFromExtProc(ctx context.Context, req *extproc.ProcessingRequest_RequestHeaders) (*http.Request, error) {
+	headers, pseudoHeaders := partitionPeusdoHeaders(req.RequestHeaders.GetHeaders().GetHeaders())
+	if err := checkPseudoRequestHeaders(pseudoHeaders); err != nil {
+		return nil, err
+	}
+
+	parsedURL, err := url.Parse(fmt.Sprintf("%s://%s%s", pseudoHeaders[":scheme"], pseudoHeaders[":authority"], pseudoHeaders[":path"]))
+	if err != nil {
+		return nil, fmt.Errorf(
+			"error building envoy URI from scheme %q, from host %q and from path %q: %w",
+			pseudoHeaders[":scheme"],
+			pseudoHeaders[":host"],
+			pseudoHeaders[":path"],
+			err)
+	}
+
+	var remoteAddr string
+	md, ok := metadata.FromIncomingContext(ctx)
+	if ok {
+		remoteAddr = getRemoteAddr(md)
+	}
+
+	var tlsState *tls.ConnectionState
+	if pseudoHeaders[":scheme"] == "https" {
+		tlsState = &tls.ConnectionState{}
+	}
+
+	headers["Host"] = append(headers["Host"], pseudoHeaders[":authority"])
+
+	return (&http.Request{
+		Method:     pseudoHeaders[":method"],
+		Host:       pseudoHeaders[":authority"],
+		RequestURI: pseudoHeaders[":path"],
+		URL:        parsedURL,
+		Header:     headers,
+		RemoteAddr: remoteAddr,
+		TLS:        tlsState,
+	}).WithContext(ctx), nil
+}
+
+type FakeResponseWriter struct {
```

Comment:
```suggestion
type fakeResponseWriter struct {
```

Reply:
Applied in db73a1f13bf8855a9c1a351a685f38629d5926bc thanks 😄